File: 01-programming-languages/C#/02-oop/02-inheritance/inheritance.cs
// C# Inheritance

using System;

namespace InheritanceDemo
{
    // Base class
    public class Animal
    {
        public string Name { get; set; }
        public int Age { get; set; }
        
        public Animal(string name, int age)
        {
            Name = name;
            Age = age;
        }
        
        public virtual string Speak()
        {
            return $"{Name} makes a sound";
        }
        
        public override string ToString()
        {
            return $"Animal: {Name}, Age: {Age}";
        }
    }
    
    // Derived class
    public class Dog : Animal
    {
        public string Breed { get; set; }
        
        public Dog(string name, int age, string breed) 
            : base(name, age)
        {
            Breed = breed;
        }
        
        public override string Speak()
        {
            return $"{Name} barks";
        }
        
        public void Fetch(string item)
        {
            Console.WriteLine($"{Name} fetches the {item}");
        }
    }
    
    // Another derived class
    public class Cat : Animal
    {
        public bool IsIndoor { get; set; }
        
        public Cat(string name, int age, bool isIndoor) 
            : base(name, age)
        {
            IsIndoor = isIndoor;
        }
        
        public override string Speak()
        {
            return $"{Name} meows";
        }
        
        public void Purr()
        {
            Console.WriteLine($"{Name} purrs...");
        }
    }
    
    // Multi-level inheritance
    public class Puppy : Dog
    {
        public bool IsTrained { get; set; }
        
        public Puppy(string name, int age, string breed, bool isTrained) 
            : base(name, age, breed)
        {
            IsTrained = isTrained;
        }
        
        public override string Speak()
        {
            return $"{Name} yips";
        }
    }
    
    // Abstract base class
    public abstract class Vehicle
    {
        public string Make { get; set; }
        public string Model { get; set; }
        public int Year { get; set; }
        
        protected Vehicle(string make, string model, int year)
        {
            Make = make;
            Model = model;
            Year = year;
        }
        
        public abstract double CalculateFuelEfficiency();
        
        public virtual string GetDescription()
        {
            return $"{Year} {Make} {Model}";
        }
    }
    
    public class Car : Vehicle
    {
        public int NumDoors { get; set; }
        
        public Car(string make, string model, int year, int numDoors) 
            : base(make, model, year)
        {
            NumDoors = numDoors;
        }
        
        public override double CalculateFuelEfficiency()
        {
            return 30.5; // MPG
        }
        
        public override string GetDescription()
        {
            return $"{base.GetDescription()} with {NumDoors} doors";
        }
    }
    
    class Program
    {
        static void Main(string[] args)
        {
            // Create objects
            Dog dog = new Dog("Rex", 5, "German Shepherd");
            Cat cat = new Cat("Whiskers", 3, true);
            Puppy puppy = new Puppy("Buddy", 1, "Labrador", false);
            
            // Method calls
            Console.WriteLine(dog.Speak());
            Console.WriteLine(cat.Speak());
            Console.WriteLine(puppy.Speak());
            
            // Polymorphism
            Animal[] animals = { dog, cat, puppy };
            foreach (Animal animal in animals)
            {
                Console.WriteLine(animal.Speak());
            }
            
            // Type checking
            if (dog is Dog d)
            {
                d.Fetch("ball");
            }
            
            // Base class reference
            Animal animalRef = dog;
            Console.WriteLine(animalRef.Speak());
            
            // Virtual method
            Console.WriteLine(dog.ToString());
            
            // Vehicle example
            Car car = new Car("Toyota", "Camry", 2023, 4);
            Console.WriteLine(car.GetDescription());
            Console.WriteLine($"Fuel efficiency: {car.CalculateFuelEfficiency():F1} MPG");
            
            Console.WriteLine("Inheritance example running");
        }
    }
}

File: 01-programming-languages/Python/02-oop/02-inheritance/inheritance.py
"""Single, multiple inheritance, and MRO."""

# ── Single Inheritance ───────────────────────────────────────────────
class Animal:
    def __init__(self, name, sound):
        self.name = name
        self.sound = sound

    def speak(self):
        return f"{self.name} says {self.sound}!"

    def __repr__(self):
        return f"Animal('{self.name}', '{self.sound}')"

class Dog(Animal):
    def __init__(self, name, breed):
        super().__init__(name, sound="Woof")  # Call parent __init__
        self.breed = breed

    def fetch(self, item):
        return f"{self.name} fetches the {item}!"

class Cat(Animal):
    def __init__(self, name, indoor=True):
        super().__init__(name, sound="Meow")
        self.indoor = indoor

    def speak(self):  # Override parent method
        return f"{self.name} purrs!"

dog = Dog("Rex", "German Shepherd")
print(dog.speak())      # Rex says Woof! (inherited)
print(dog.fetch("ball"))  # Rex fetches the ball!
print(dog.breed)        # German Shepherd

cat = Cat("Whiskers")
print(cat.speak())      # Whiskers purrs! (overridden)

# ── Multiple Inheritance ─────────────────────────────────────────────
class Swimmer:
    def swim(self):
        return f"{self.name} is swimming"

class Flyer:
    def fly(self):
        return f"{self.name} is flying"

class Duck(Animal, Swimmer, Flyer):
    def __init__(self, name):
        super().__init__(name, sound="Quack")

donald = Duck("Donald")
print(donald.speak())    # Donald says Quack!
print(donald.swim())    # Donald is swimming
print(donald.fly())     # Donald is flying

# ── Method Resolution Order (MRO) ───────────────────────────────────
# Python uses C3 linearization to determine method lookup order
print(Duck.__mro__)
# (Duck, Animal, Swimmer, Flyer, object)

# super() follows MRO — calls next class in chain
class A:
    def greet(self):
        return "A"

class B(A):
    def greet(self):
        return "B" + super().greet()

class C(A):
    def greet(self):
        return "C" + super().greet()

class D(B, C):
    def greet(self):
        return "D" + super().greet()

print(D().greet())  # DCBA
print(D.__mro__)    # D, B, C, A, object

# ── isinstance and issubclass ────────────────────────────────────────
print(isinstance(dog, Dog))         # True
print(isinstance(dog, Animal))      # True
print(isinstance(dog, (Dog, Cat)))  # True (tuple check)

print(issubclass(Dog, Animal))      # True
print(issubclass(Dog, object))      # True

# ── Composition Over Inheritance ────────────────────────────────────
class Engine:
    def start(self):
        return "Engine started"

class Car:
    def __init__(self, engine):
        self.engine = engine  # Composition — has-a relationship

    def start(self):
        return self.engine.start()

car = Car(Engine())
print(car.start())  # Engine started

# ── Abstract Base Class for Interface ───────────────────────────────
from abc import ABC, abstractmethod

class Shape(ABC):
    @abstractmethod
    def area(self):
        pass

    @abstractmethod
    def perimeter(self):
        pass

class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius

    def area(self):
        import math
        return math.pi * self.radius ** 2

    def perimeter(self):
        import math
        return 2 * math.pi * self.radius
